### PR TITLE
Enable strict null checks.

### PR DESCRIPTION
**Pull request by rictic**
Also, put an array of warnings on all features and scanned features so
that instead of bailing out or throwing we can instead shove a warning
someplace that may be useful in future.

Was otherwise rather straightforward. Caught several bugs. Should have done this a long time
ago I think.

### CODE REVIEW

File: src/editor-service/editor-server.ts [hunk 1]
```diff
@@ -97,7 +97,7 @@ process.stdin.pipe(split()).on('data', async function(line: string) {
 // node child_process.fork() IPC interface
 process.on('message', async function(request: RequestWrapper) {
   const result = await getSettledValue(request.value);
-  process.send(<ResponseWrapper>{id: request.id, value: result});
+  process.send!(<ResponseWrapper>{id: request.id, value: result});
```
Comment (FredKSchott) [inline src/editor-service/editor-server.ts]: can you remind me (any anyone else reviewing) what `!` does when used like this?

File: src/editor-service/local-editor-service.ts [hunk 2]
```diff
@@ -80,7 +85,7 @@ export class LocalEditorService extends EditorService {
       let attributes: AttributeCompletion[] = [];
       for (const element of elements) {
         // A map from the inheritedFrom to a sort prefix.
-        let sortPrefixes = new Map<string, string>();
+        let sortPrefixes = new Map<string|undefined|null, string>();
```
Comment (FredKSchott) [inline src/editor-service/local-editor-service.ts]: can we get rid of undefined and null as valid keys into this map?

File: src/generate-elements.ts [hunk 3]
```diff
@@ -85,7 +86,7 @@ function serializeElement(
   const events = resolvedElement.events.map(
       e => ({
         name: e.name,
-        description: e.description,
+        description: e.description || '',
```
Comment (FredKSchott) [inline src/generate-elements.ts]: can we leave as undefined if no description is found? It would make detection easier and less error prone downstream.

File: src/html/html-script-scanner.ts [hunk 4]
```diff
@@ -43,16 +43,16 @@ export class HtmlScriptScanner implements HtmlScanner {
...
           const importUrl = resolveUrl(document.url, src);
           features.push(new ScannedScriptTagImport(
-              'html-script', importUrl, document.sourceRangeForNode(node),
-              document.sourceRangeForAttribute(node, 'src'), node));
+              'html-script', importUrl, document.sourceRangeForNode(node)!,
+              document.sourceRangeForAttribute(node, 'src')!, node));
         } else {
           const locationOffset = getLocationOffsetOfStartOfTextContent(node);
-          const attachedCommentText = getAttachedCommentText(node);
+          const attachedCommentText = getAttachedCommentText(node) || '';
```
Comment (FredKSchott) [inline src/html/html-script-scanner.ts]: same here and for more below, if there's no attached comment text can we leave as undefined?

File: src/javascript/esutil.ts [hunk 5]
```diff
@@ -157,12 +165,24 @@ export function toScannedPolymerProperty(
   let description =
       jsdoc.removeLeadingAsterisks(getAttachedComment(node) || '').trim();
 
+  const name = objectKeyToString(node.key);
+  if (name == null) {
```
Comment (FredKSchott) [inline src/javascript/esutil.ts]: this will not be null anymore, use `typeof === 'undefined'` if possible.

File: src/javascript/javascript-document.ts [hunk 6]
```diff
@@ -23,7 +23,7 @@ import {Visitor, VisitResult} from './estree-visitor';
 
 export {Visitor} from './estree-visitor';
 
-let __exampleNode: Node;
+const __exampleNode: Node = <any>null;
```
Comment (FredKSchott) [inline src/javascript/javascript-document.ts]: can you add a comment explaining this? Curious about its usage here

File: src/model/element.ts [hunk 7]
```diff
@@ -32,13 +33,14 @@ export class ScannedElement implements Resolvable {
   className?: string;
   superClass?: string;
   extends?: string;
-  properties: ScannedProperty[] = [];
-  attributes: ScannedAttribute[] = [];
+  properties = [] as ScannedProperty[];
```
Comment (FredKSchott) [inline src/model/element.ts]: why the change to the new format?

File: src/editor-service/editor-server.ts (hunk 1 above)
Comment (rictic) [inline src/editor-service/editor-server.ts]: It means "I solemnly swear the expression to the left of the `!` is never null or undefined (right here just this one time)"

File: src/html/html-script-scanner.ts (hunk 4 above)
Comment (rictic) [inline src/html/html-script-scanner.ts]: see comment above

File: src/editor-service/local-editor-service.ts (hunk 2 above)
Comment (FredKSchott) [inline src/editor-service/local-editor-service.ts]: Got it, that makes sense. Can you document this somewhere?

File: src/javascript/esutil.ts (hunk 5 above)
Comment (FredKSchott) [inline src/javascript/esutil.ts]: Oh right, I remember we've talked about this before. Using `(name == null)` to check for falseyness is deceptive to me as a reviewer, because it feels like the intention is to check something more specific about the variable & `null` and maybe the `==` was a mistake.  `(!name)` more explicitly checks for falseyness.

Can we switch this to `(!name)` here and going forward?

File: src/model/element.ts (hunk 7 above)
Comment (FredKSchott) [inline src/model/element.ts]: I like that you're experimenting! :)

I personally prefer the first form (matches non default property syntax & uses types to describe the property instead of the value being set to that property) but I guess that's subjective. I'm more concerned with moving away from the format we've been using everywhere else so far and would prefer we stick to that.

File: src/generate-elements.ts (hunk 3 above)
Comment (FredKSchott) [inline src/generate-elements.ts]: Got it, thanks for the explanation. I still feel like `undefined` is more true here and will be an easier interface for users to interact with, but I do see the value in simplifying our types and elements.json format. I'll leave the decision up to you :)

File: src/javascript/esutil.ts (hunk 5 above)
Comment (rictic) [inline src/javascript/esutil.ts]: Here, I think you're totally right, `!name` is better, as an empty property name doesn't make sense as a polymer property.

In general I think we should keep `foo == null` in our repertoire, as it only catches `null` and `undefined` whereas `!foo` also catches `false`, `0`, `''`, `NaN`, (and in the browser, `document.all` because lol), and many times you do want to distinguish between nullish and falsy.

File: src/model/element.ts (hunk 7 above)
Comment (rictic) [inline src/model/element.ts]: @justinfagnani care to weigh in on a question of style?

File: src/javascript/javascript-document.ts (hunk 6 above)
Comment (rictic) [inline src/javascript/javascript-document.ts]: Added a comment. 

(suddenly I can reply here. still figuring out github's new review ui..)

File: src/editor-service/editor-server.ts (hunk 1 above)
Comment (justinfagnani) [inline src/editor-service/editor-server.ts]: It's a cast. If you have `o: T|null`, `o!` is the same as `(o as T)`. If TypeScript had subtraction types, it'd be `(o as (typeof o - null|undefined))`.

File: src/javascript/esutil.ts (hunk 5 above)
Comment (FredKSchott) [inline src/javascript/esutil.ts]: wait... `false != null`? How did I forget that???

Nevermind then, there's definitely value in distinguishing between the two

File: src/generate-elements.ts (hunk 3 above)
Comment (justinfagnani) [inline src/generate-elements.ts]: Before null-tracking I would say that I prefer non-null fields so that client code doesn't have to branch too many times to deal with null. I still think this is true, especially with collections. If the primary client branching would come from detecting empty though, it might not buy much, and with null-tracking you at least get warnings.

So `null` could make sense here. I'm not so sure about `undefined` as the 'not present' value. Here I'd like to represent that the object has a description property, that may be null, rather than not having a description property at all. Of course, to be maximally confusing, setting a property to `undefined` in JS means that the object has an own property name, but it definitely clouds the issue.

**Comment by justinfagnani** [comment]
Before I get too far in this, what do we think about the default 'no value' value being `null`, not `undefined`?

File: src/model/element.ts (hunk 7 above)
Comment (justinfagnani) [inline src/model/element.ts]: I think it's a toss-up here. One is a type annotation to fill in a gap in type inference, the other is a cast, which then makes type inference work. Here, I think I prefer the type annotation so that the list of variables and their types is relatively consistent. It's also good to have a consistent `let or const` `name` `type` `initializer` ordering.

File: src/generate-elements.ts (hunk 3 above)
Comment (rictic) [inline src/generate-elements.ts]: Client code for elements.json is potentially really diverse, it's not just nodejs and not just code that we write, so I wanted to make it as easy to consume as possible. IMO defaulting to a description of empty string is the easiest thing for a user to consume.

Agreed on generally preferring nulls to undefined to represent a 'not present' value on a field.

File: src/analyzer.ts [hunk 8]
```diff
@@ -269,8 +264,8 @@ export class Analyzer {
           }
         });
 
-    const dependencies =
-        (await Promise.all(scannedSubDocuments)).filter(s => !!s);
+    const dependencies = (await Promise.all(scannedSubDocuments))
+                             .filter(s => !!s) as ScannedDocument[];
```
Comment (rictic) [inline src/analyzer.ts]: Not without sacrificing local correctness, because we're only filtering out the nulls here.

File: src/model/element.ts [hunk 9]
```diff
@@ -70,6 +72,7 @@ export class Element implements Feature {
   jsdoc?: jsdoc.Annotation;
   astNode: estree.Node|null;
   kinds: Set<string> = new Set(['element']);
+  warnings: Warning[] = [];
```
Comment (FredKSchott) [inline src/model/element.ts]: What kind of feature can have warnings now? Does it make sense to add directly to the Feature class? Then all documents and/or elements could to have a `getWarnings()` method that returns all warnings contained within.

WDYT? I've been thinking about how to return all warnings found within a file, which is why this is top of mind.

Comment (rictic) [inline src/model/element.ts]: Feature is just an interface so we need to add this onto all implementing classes, but yes. This PR adds an array of warnings onto ~everything, for getWarnings to gather up

**Comment by rictic** [comment]
Rebased and comments addressed, PTAL

**Comment by justinfagnani** [comment]
LGTM
